Sprint 3/step 1

Nikba-Creative-Studio opened the pull request:
### Проект. Роутинг в React приложении. Реализация механизмов авторизации в React приложении.

Деплой: https://demo5.nikba-dev.com/

**Список маршрутов**

- / - главная страница, конструктор бургеров.
- /login - страница авторизации.
- /register - страница регистрации.
- /forgot-password - страница восстановления пароля.
- /reset-password - страница сброса пароля.
- /profile — страница с настройками профиля пользователя.
- /ingredients/:id — страница ингредиента. 

1. Страницы авторизации и регистрации
2. Страницы восстановления и сброса пароля
3. Страница профиля пользователя
4. Авторизация и регистрация
5. Получение и обновление информации о пользователе
6. Защищённые маршруты в приложении
7. Доработка роутинга в приложении
8. Ручное тестирование роутинга
9. Проверка типизации

Nikba-Creative-Studio committed:
1. Страницы авторизации и регистрации - Вверстка

Nikba-Creative-Studio committed:
Регистрация

Nikba-Creative-Studio committed:
Авторизация

Nikba-Creative-Studio committed:
Куки

Nikba-Creative-Studio committed:
Страницы 404

Nikba-Creative-Studio committed:
Авторизация, Обновление токена, Выход из системы

Nikba-Creative-Studio committed:
Получение и обновление информации о пользователе

Nikba-Creative-Studio committed:
Обновление токена

Nikba-Creative-Studio committed:
Доработка роутинга в приложении

Nikba-Creative-Studio committed:
Ручное тестирование роутинга

irinamozes commented:
Проект очень неплохой.

Правильно показывается модальное окно с деталями ингредиента и его детальная страница.

Верно организован роутинг для авторизованного и неавторизованного пользователей.

Что нужно исправить.

1. Вы посылаете запрос на получение одного и того же массива ингредиентов в компонентах App и IngredientDetails. Но, поскольку это один и тот же массив, запрос на его получение должен происходить один раз в проекте, в компоненте App при монтировании сайта, чтобы массив ингредиентов был доступен всем страницам без повторения запроса на получение ингредиентов на этих страницах.
Может случиться так, что детальная страница ингредиента уже загрузилась и должна показываться, а массив ингредиентов, который на эту страницу должен передаваться по useSelector, ещё не загрузился в App. Поэтому, прежде чем делать рендер на детальнойя странице ингредиента, нужно наряду с проверкой наличия массива ингредиентов (нужно проверять равна или не равна его длина 0) надо проверять значение переменных состояния типа isLoading, isError, которые должны менять  своё значение в процессе взаимодействия с сервером при запросе ингредиентов. Эти переменные состояния нужно создать в части Redux проекта, отвечающей за получение массива ингредиентов. Также при асинхронном взаимодействии с сервером должен посылаться ещё один тип экшена, имеющий в названии типа слово _REQUEST. 
Прочитайте, пожалуйста, об асинхронных экшенах и структуре части хранилища, отвечающей за API, в лекциях на платформе с тренажёрами для 2-го этапа 2-го спринта.

2. При запросе пользователя на ендпоинт user при просроченности аксесс-токена (по истечении 20 минут после авторизации) этот запрос должен автоматически повторяться после автоматического же обновления токенов.
Сейчас этого не происходит. При попытке сохранения данных на странице профиля в этой ситуации выдаётся сообщение, что произошла ошибка и данные не сохраняются (см. снимок).
![Снимок экрана в 2022-02-28 18-34-13](https://user-images.githubusercontent.com/19558891/156010436-dbd626ac-9e23-4c94-bcaf-edb603819b05.png)
Напоминаю, как должен быть реализован данный функционал.
Аксесс  токен предназначен только для авторизованного пользователя и его валидность проверяется сервером при запросах авторизованного пользователя к серверу за конфиденциальной информацией пользователя. То есть, нужно отслеживать сообщение сервера 'jwt expired'(или статус 403 ошибки в ответе сервера) при запросах на получение данных профиля и при обновлении данных профиля.
При сообщении сервера 'jwt expired' при запросах на получение данных профиля и при их обновлении, нужно посылать запрос на ендпоинт token для обновления аксесс-токена в блоках catch методов с этими запросами, а затем повторять запрос на получение данных профиля или обновление данных профиля, которые не удались перед этим, в этих же блоках catch  и с обновлённым аксесс-токеном. Таким образом, при просроченности аксесс-токена обновление токенов для пользователя должно происходить незаметно, и также незаметно должен повторяться неудавшийся запрос с обновлёнными токенами.
По смыслу аксесс-токена отслежиать сообщение сервера 'jwt expired' нужно  и при оформлении заказа авторизованным пользователем.
Смысл аксесс и рефреш токенов и схема работы с ними на клиенте описаны здесь: https://ru.wikipedia.org/wiki/JSON_Web_Token.

3. Мне почему-то не пришло письмо на почту с кодом для восстановления пароля, который нужно вставлять на странице /reset-password. Возможно не сработал сервер приложения от практикума, но на всякий случай проверьте правильность этого функционала.

Nikba-Creative-Studio committed:
Правки

1. Запрос на получение массива ингредиентов остался только один в App. Также реализован функционал спиннера загрузки.
2. На функций
 - getUser - получение данных пользователя
 - editUser - изменения пользователя
 - postOrder - отправка заказа на сервер

в catch при статусе 403 запускается функция updateToken

postOrder - отправка заказа на сервер делается через авторизацию, как результат видим что данный заказ привязан к определенному пользователю.

Nikba-Creative-Studio committed:
Переход с 403 на jwt expired

Nikba-Creative-Studio committed:
Loading для отправки заказа

irinamozes commented:
Проект отличный!
Все требования задания выполнены.

Спустя 20 минут после входа в ЛК при сохранении измененённых данных профиля обновились токены, как и должно быть, и неудавшийся до этого запрос на обновление данных профиля повторился. Обновлённые данные сохранились (см. снимок).
![Снимок экрана в 2022-03-01 22-51-14](https://user-images.githubusercontent.com/19558891/156225739-f7a81298-290e-4119-9342-84ada9021a0c.png)

Не забудьте в задании по созданию ленты заказов учитывать асинхронность работы сервера в компонентах, которые должны выводить на экран размётку на основе ответа сервера. Это поможет избежать ошибок в консоли, происходящих из-за попыток что-то рендерить, когда сервер ещё не успел прислать данные для рендера, или произошла ошибка при запросе к серверу, или сервер вернул статус успеха, но не вернул реальных данных.

Спринт принимается.

Желаю дальнейших успехов в учёбе и удачи!
